crypto/x509/v3_asid.c

```diff
@@ -196,18 +199,23 @@ int X509v3_asid_add_id_or_range(ASIdentifiers *asid,
     default:
         return 0;
     }
-    if (*choice != NULL && (*choice)->type == ASIdentifierChoice_inherit)
+    if (*choice != NULL && (*choice)->type != ASIdentifierChoice_asIdsOrRanges)
         return 0;
     if (*choice == NULL) {
         if ((*choice = ASIdentifierChoice_new()) == NULL)
             return 0;
         (*choice)->u.asIdsOrRanges = sk_ASIdOrRange_new(ASIdOrRange_cmp);
-        if ((*choice)->u.asIdsOrRanges == NULL)
+        if ((*choice)->u.asIdsOrRanges == NULL) {
+            ASIdentifierChoice_free(*choice);
+            *choice = NULL;
             return 0;
+        }
         (*choice)->type = ASIdentifierChoice_asIdsOrRanges;
     }
     if ((aor = ASIdOrRange_new()) == NULL)
         return 0;
+    if (!sk_ASIdOrRange_reserve((*choice)->u.asIdsOrRanges, 1))
+        goto err;
```

Comment:
Once you called `ASIdentifierChoice_new()`, you have created an `asid` that will no longer be a valid `ASIdentifier` in many contexts. In particular, its DER encoding will be invalid until you actually attach an ASId or an ASRange.

Reply:
We only want this function to fail safe after a memory allocation error.
The only reasonable thing that the caller can do in the event
of an unexpected memory allocation error, is abort whatever is
going on, and clean up everything.
The application can expect not to create a memory leak of a crash
due to bugs in the error handling.  I have an error injection test
running that detected this error and many others.